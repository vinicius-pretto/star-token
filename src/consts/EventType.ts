enum EventType {
  GET_TOKENS = "GET_TOKENS",
  SET_TOKENS = "SET_TOKENS",
  SET_COLOR_TOKEN = "SET_COLOR_TOKEN",
  SET_FONT_SIZE_TOKEN = "SET_FONT_SIZE_TOKEN",
  SET_BORDER_RADIUS_TOKEN = "SET_BORDER_RADIUS_TOKEN",
  UPDATE_COLOR_TOKEN = "UPDATE_COLOR_TOKEN",
  UPDATE_FONT_SIZE_TOKEN = "UPDATE_FONT_SIZE_TOKEN",
  UPDATE_BORDER_RADIUS_TOKEN = "UPDATE_BORDER_RADIUS_TOKEN",
  DELETE_TOKEN = "DELETE_TOKEN",
}

export default EventType;
