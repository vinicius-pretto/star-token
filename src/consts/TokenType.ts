enum TokenType {
  COLOR = "color",
  FONT_SIZE = "font_size",
  BORDER_RADIUS = "border_radius",
}

export default TokenType;
