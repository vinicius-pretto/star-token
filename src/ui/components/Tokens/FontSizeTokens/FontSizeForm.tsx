import * as React from "react";
import Input from "../../Input/Input";
import Modal from "../../Modal/Modal";

const FontSizeForm = ({ isModalOpen, onCloseModal, formik }) => {
  const onValueChange = (e) => {
    const input = e.target.value;
    const inputFiltered = input.replace(/[^0-9]/g, "");
    e.target.value = inputFiltered;
    formik.handleChange(e);
  };

  return (
    <Modal title="Font Size" isOpen={isModalOpen} onClose={onCloseModal}>
      <form onSubmit={formik.handleSubmit}>
        <Input
          id="name"
          type="text"
          value={formik.values.name}
          onChange={formik.handleChange}
          error={formik.errors.name}
          touched={formik.touched.name}
          placeholder="font-size-base"
        />

        <Input
          id="value"
          type="text"
          maxLength={6}
          value={formik.values.value}
          onChange={onValueChange}
          error={formik.errors.value}
          touched={formik.touched.value}
          placeholder="16"
        />

        <div className="d-flex justify-content-end">
          <button type="submit" className="button button--primary">
            Save
          </button>
        </div>
      </form>
    </Modal>
  );
};

export default FontSizeForm;
