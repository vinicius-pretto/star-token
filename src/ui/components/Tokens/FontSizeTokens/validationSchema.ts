import * as yup from "yup";

const validationSchema = yup.object().shape({
  name: yup
    .string()
    .required("Required")
    .max(48, "Maximum 48 characters")
    .matches(
      /^[a-zA-Z]{1,1}[a-zA-Z0-9\-\_]*$/,
      "Must start with a letter and contain only letters, numbers, hyphens (-), and underscores (_)"
    ),
  value: yup.string().required("Required"),
});

export default validationSchema;
