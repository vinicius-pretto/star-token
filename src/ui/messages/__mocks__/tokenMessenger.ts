const tokenMessenger = {
  postGetTokensMessage: jest.fn(),
  postSetTokensMessage: jest.fn(),
  postUpdateColorTokenMessage: jest.fn(),
  postDeleteTokenMessage: jest.fn(),
  postMessage: jest.fn(),
};

export default tokenMessenger;
