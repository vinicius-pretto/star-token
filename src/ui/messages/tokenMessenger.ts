import EventType from "../../consts/EventType";
import Token from "../../consts/Token";

const TARGET_ORIGIN = "*";

const postGetTokensMessage = () => {
  const message = {
    pluginMessage: {
      type: EventType.GET_TOKENS,
    },
  };
  window.parent.postMessage(message, TARGET_ORIGIN);
};

const postSetTokensMessage = (tokens) => {
  const message = {
    pluginMessage: {
      type: EventType.SET_TOKENS,
      tokens,
    },
  };
  window.parent.postMessage(message, TARGET_ORIGIN);
};

const postSetColorTokenMessage = (token: Token) => {
  const message = {
    pluginMessage: {
      type: EventType.SET_COLOR_TOKEN,
      token,
    },
  };
  window.parent.postMessage(message, TARGET_ORIGIN);
};

const postMessage = ({ type, payload }) => {
  const message = {
    pluginMessage: {
      type,
      payload,
    },
  };
  window.parent.postMessage(message, TARGET_ORIGIN);
};

const postUpdateColorTokenMessage = (token: Token) => {
  const message = {
    pluginMessage: {
      type: EventType.UPDATE_COLOR_TOKEN,
      token,
    },
  };
  window.parent.postMessage(message, TARGET_ORIGIN);
};

const postDeleteTokenMessage = (tokenId: string) => {
  const message = {
    pluginMessage: {
      type: EventType.DELETE_TOKEN,
      tokenId,
    },
  };
  window.parent.postMessage(message, TARGET_ORIGIN);
};

export default {
  postGetTokensMessage,
  postSetTokensMessage,
  postSetColorTokenMessage,
  postUpdateColorTokenMessage,
  postDeleteTokenMessage,
  postMessage,
};
