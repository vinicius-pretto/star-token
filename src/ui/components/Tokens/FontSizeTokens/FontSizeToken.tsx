import * as React from "react";
import EventType from "../../../../consts/EventType";
import tokenMessenger from "../../../messages/tokenMessenger";

const FontSizeToken = ({ token, onUpdate, onDelete }) => {
  const setTokenStyle = () => {
    tokenMessenger.postMessage({
      type: EventType.SET_FONT_SIZE_TOKEN,
      payload: token,
    });
  };

  return (
    <div
      key={token.id}
      className="d-flex flex-row align-items-center justify-content-between w-100 color-token-container"
    >
      <button className="color-token" onClick={setTokenStyle}>
        <span className="mr-md font-token-shape">Ag</span>
        <span>{token.name}</span>
      </button>

      <div className="d-flex flex-row">
        <div
          className="icon-button visibility-hidden"
          role="button"
          aria-label={`Edit ${token.name}`}
          tabIndex={0}
          onClick={() => onUpdate(token)}
        >
          <div className="icon icon--adjust"></div>
        </div>

        <div
          className="icon-button visibility-hidden"
          role="button"
          aria-label={`Delete ${token.name}`}
          tabIndex={0}
          onClick={() => onDelete(token.id)}
        >
          <div className="icon icon--trash"></div>
        </div>
      </div>
    </div>
  );
};

export default FontSizeToken;
