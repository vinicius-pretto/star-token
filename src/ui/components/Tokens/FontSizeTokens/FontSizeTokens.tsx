import * as React from "react";
import _isEmpty from "lodash/isEmpty";
import _cloneDeep from "lodash/cloneDeep";
import _findIndex from "lodash/findIndex";
import { v4 as uuid } from "uuid";
import { useDispatch } from "react-redux";
import Token from "../../../../consts/Token";
import FontSizeToken from "./FontSizeToken";
import FontSizeForm from "./FontSizeForm";
import { useFormik } from "formik";
import validationSchema from "./validationSchema";
import TokenType from "../../../../consts/TokenType";
import { storeTokens } from "../../../redux/actions";
import tokenMessenger from "../../../messages/tokenMessenger";
import EventType from "../../../../consts/EventType";

const FontSizeTokens = ({ tokens, onDelete }) => {
  const fontSizeTokens = tokens.filter(
    (token: Token) => token.type === TokenType.FONT_SIZE
  );
  const dispatch = useDispatch();
  const [isModalOpen, setIsModalOpen] = React.useState(false);
  const [tokenSelected, setTokenSelected] = React.useState({});
  const initialValues = {
    name: "",
    value: "",
  };

  const onSubmit = (values) => {
    if (!_isEmpty(tokenSelected)) {
      const tokenUpdated = {
        ...tokenSelected,
        name: values.name,
        value: values.value,
      };
      updateToken(tokenUpdated);
      return;
    }
    const token = {
      id: uuid(),
      type: TokenType.FONT_SIZE,
      name: values.name,
      value: values.value,
    };
    const tokensUpdated = tokens.concat(token);

    onCloseModal();
    clearFields();
    dispatch(storeTokens(tokensUpdated));
  };

  const formik = useFormik({
    initialValues,
    validationSchema,
    onSubmit,
  });

  const clearFields = () => {
    setTokenSelected({});
    formik.resetForm();
  };

  const onCreate = () => {
    clearFields();
    onOpenModal();
  };

  const onUpdateToken = (token: Token) => {
    const values = { name: token.name, value: token.value };
    onOpenModal();
    formik.setValues(values);
    setTokenSelected(token);
  };

  const updateToken = (token) => {
    const tokensCopy = _cloneDeep(tokens);
    const index = _findIndex(tokensCopy, ["id", token.id]);
    tokensCopy.splice(index, 1, token);

    dispatch(storeTokens(tokensCopy));
    clearFields();
    onCloseModal();
    tokenMessenger.postMessage({
      type: EventType.UPDATE_FONT_SIZE_TOKEN,
      payload: token,
    });
  };

  const onOpenModal = () => {
    setIsModalOpen(true);
  };

  const onCloseModal = () => {
    setIsModalOpen(false);
  };

  return (
    <section className="border-bottom p-sm">
      <div className="tokens-section-header">
        <h3 className="section-title">Font Size</h3>

        <div
          className="icon-button"
          role="button"
          aria-label="Add token"
          tabIndex={0}
          onClick={onCreate}
          data-testid="add-token-btn"
        >
          <div className="icon icon--plus"></div>
        </div>
      </div>

      {_isEmpty(fontSizeTokens) ? (
        <p className="label my-0">No font size tokens</p>
      ) : (
        <div className="d-flex flex-row flex-wrap">
          {fontSizeTokens.map((token: Token) => (
            <FontSizeToken
              key={token.id}
              token={token}
              onUpdate={onUpdateToken}
              onDelete={onDelete}
            />
          ))}
        </div>
      )}

      <FontSizeForm
        isModalOpen={isModalOpen}
        onCloseModal={onCloseModal}
        formik={formik}
      />
    </section>
  );
};

export default FontSizeTokens;
