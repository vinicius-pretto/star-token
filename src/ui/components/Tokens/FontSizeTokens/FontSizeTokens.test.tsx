import * as React from "react";
import _ from "lodash";
import userEvent from "@testing-library/user-event";
import { fireEvent } from "@testing-library/dom";
import { render, screen, waitFor } from "../../../../testUtils";
import FontSizeTokens from "./FontSizeTokens";
import tokens from "../../../../testData/tokens";
import tokenMessenger from "../../../messages/tokenMessenger";
import TokenType from "../../../../consts/TokenType";
import EventType from "../../../../consts/EventType";

jest.mock("../../../messages/tokenMessenger");

describe("FontSizeTokens", () => {
  afterEach(() => {
    jest.clearAllMocks();
  });

  describe("snapshot", () => {
    it("no color tokens", () => {
      const { asFragment } = render(
        <FontSizeTokens tokens={[]} onDelete={jest.fn()} />
      );
      expect(asFragment()).toMatchSnapshot();
    });

    it("with color tokens", () => {
      const { asFragment } = render(
        <FontSizeTokens tokens={tokens} onDelete={jest.fn()} />
      );
      expect(asFragment()).toMatchSnapshot();
    });
  });

  describe("form validation", () => {
    beforeEach(() => {
      render(<FontSizeTokens tokens={[]} onDelete={jest.fn()} />);

      const addButton = screen.getByTestId("add-token-btn");
      userEvent.click(addButton);
    });

    it("display error when submit empty form", async () => {
      const submitButton = screen.getByRole("button", { name: /save/i });
      await waitFor(() => userEvent.click(submitButton));

      expect(screen.getAllByText("Required")).toHaveLength(2);
      expect(tokenMessenger.postSetTokensMessage).not.toHaveBeenCalled();
    });

    describe("input name", () => {
      it("display error when fill out invalid character", async () => {
        const inputName = screen.getByPlaceholderText("font-size-base");
        const submitButton = screen.getByRole("button", { name: /save/i });
        const errorMessage =
          "Must start with a letter and contain only letters, numbers, hyphens (-), and underscores (_)";

        userEvent.type(inputName, "-/*+");
        await waitFor(() => userEvent.click(submitButton));

        expect(screen.getByText(errorMessage)).toBeInTheDocument();
      });

      it("display error when length is greather than 48 characters", async () => {
        const inputName = screen.getByPlaceholderText("font-size-base");
        const submitButton = screen.getByRole("button", { name: /save/i });
        const errorMessage = "Maximum 48 characters";

        userEvent.type(inputName, "a".repeat(49));
        await waitFor(() => userEvent.click(submitButton));

        expect(screen.getByText(errorMessage)).toBeInTheDocument();
      });
    });

    describe("input value", () => {
      it("replace characters which does not follow the pattern A-Z|0-9", async () => {
        const inputName = screen.getByPlaceholderText("font-size-base");
        const inputValue = screen.getByPlaceholderText("16");
        const submitButton = screen.getByRole("button", { name: /save/i });

        userEvent.type(inputName, "font-size-base");
        userEvent.type(inputValue, "/*aZ*/-)$!%16");
        await waitFor(() => userEvent.click(submitButton));

        expect(tokenMessenger.postSetTokensMessage).toHaveBeenCalledTimes(1);
        expect(tokenMessenger.postSetTokensMessage).toHaveBeenCalledWith([
          {
            id: expect.any(String),
            name: "font-size-base",
            type: TokenType.FONT_SIZE,
            value: "16",
          },
        ]);
      });
    });
  });

  it("update token", async () => {
    render(<FontSizeTokens tokens={tokens} onDelete={jest.fn()} />);

    const editButton = screen.getByRole("button", {
      name: /edit font-size-base/i,
    });

    userEvent.hover(screen.getByText("font-size-base"));
    await waitFor(() => userEvent.click(editButton));

    const inputName = screen.getByPlaceholderText("font-size-base");
    const inputValue = screen.getByPlaceholderText("16");
    const submitButton = screen.getByRole("button", { name: /save/i });

    expect(inputName).toHaveValue("font-size-base");
    expect(inputValue).toHaveValue("16");

    await waitFor(() =>
      fireEvent.change(inputName, { target: { value: "font-size-md" } })
    );
    await waitFor(() =>
      fireEvent.change(inputValue, { target: { value: "18" } })
    );
    await waitFor(() => userEvent.click(submitButton));

    const tokensCopy = _.cloneDeep(tokens);
    const index = _.findIndex(tokensCopy, [
      "id",
      "6552cefb-8e70-43ba-b6c7-432a8b8aa591",
    ]);
    const tokenUpdated = {
      id: "6552cefb-8e70-43ba-b6c7-432a8b8aa591",
      type: TokenType.FONT_SIZE,
      name: "font-size-md",
      value: "18",
    };
    tokensCopy.splice(index, 1, tokenUpdated);

    expect(tokenMessenger.postSetTokensMessage).toHaveBeenCalledTimes(1);
    expect(tokenMessenger.postSetTokensMessage).toHaveBeenCalledWith(
      tokensCopy
    );
    expect(tokenMessenger.postMessage).toHaveBeenCalledTimes(1);
    expect(tokenMessenger.postMessage).toHaveBeenCalledWith({
      type: EventType.UPDATE_FONT_SIZE_TOKEN,
      payload: tokenUpdated,
    });
  });
});
