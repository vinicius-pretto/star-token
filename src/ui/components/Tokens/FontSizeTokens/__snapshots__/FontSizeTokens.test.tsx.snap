// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FontSizeTokens snapshot no color tokens 1`] = `
<DocumentFragment>
  <section
    class="border-bottom p-sm"
  >
    <div
      class="tokens-section-header"
    >
      <h3
        class="section-title"
      >
        Font Size
      </h3>
      <div
        aria-label="Add token"
        class="icon-button"
        data-testid="add-token-btn"
        role="button"
        tabindex="0"
      >
        <div
          class="icon icon--plus"
        />
      </div>
    </div>
    <p
      class="label my-0"
    >
      No font size tokens
    </p>
    <div
      class=""
      id="modal"
    >
      <div
        class="modal-dialog"
      >
        <div
          class="modal-dialog-header"
        >
          <h2
            class="section-title"
          >
            Font Size
          </h2>
          <div
            aria-label="Close"
            class="icon-button"
            role="button"
            tabindex="0"
          >
            <div
              class="icon icon--close"
            />
          </div>
        </div>
        <div
          class="modal-dialog-body"
        >
          <form>
            <div
              class="input mb-sm form-group"
            >
              <input
                class="input__field"
                id="name"
                placeholder="font-size-base"
                type="text"
                value=""
              />
              <p
                class="visibility-hidden"
              />
            </div>
            <div
              class="input mb-sm form-group"
            >
              <input
                class="input__field"
                id="value"
                maxlength="6"
                placeholder="16"
                type="text"
                value=""
              />
              <p
                class="visibility-hidden"
              />
            </div>
            <div
              class="d-flex justify-content-end"
            >
              <button
                class="button button--primary"
                type="submit"
              >
                Save
              </button>
            </div>
          </form>
        </div>
      </div>
    </div>
  </section>
</DocumentFragment>
`;

exports[`FontSizeTokens snapshot with color tokens 1`] = `
<DocumentFragment>
  <section
    class="border-bottom p-sm"
  >
    <div
      class="tokens-section-header"
    >
      <h3
        class="section-title"
      >
        Font Size
      </h3>
      <div
        aria-label="Add token"
        class="icon-button"
        data-testid="add-token-btn"
        role="button"
        tabindex="0"
      >
        <div
          class="icon icon--plus"
        />
      </div>
    </div>
    <div
      class="d-flex flex-row flex-wrap"
    >
      <div
        class="d-flex flex-row align-items-center justify-content-between w-100 color-token-container"
      >
        <button
          class="color-token"
        >
          <span
            class="mr-md font-token-shape"
          >
            Ag
          </span>
          <span>
            font-size-base
          </span>
        </button>
        <div
          class="d-flex flex-row"
        >
          <div
            aria-label="Edit font-size-base"
            class="icon-button visibility-hidden"
            role="button"
            tabindex="0"
          >
            <div
              class="icon icon--adjust"
            />
          </div>
          <div
            aria-label="Delete font-size-base"
            class="icon-button visibility-hidden"
            role="button"
            tabindex="0"
          >
            <div
              class="icon icon--trash"
            />
          </div>
        </div>
      </div>
      <div
        class="d-flex flex-row align-items-center justify-content-between w-100 color-token-container"
      >
        <button
          class="color-token"
        >
          <span
            class="mr-md font-token-shape"
          >
            Ag
          </span>
          <span>
            font-size-lg
          </span>
        </button>
        <div
          class="d-flex flex-row"
        >
          <div
            aria-label="Edit font-size-lg"
            class="icon-button visibility-hidden"
            role="button"
            tabindex="0"
          >
            <div
              class="icon icon--adjust"
            />
          </div>
          <div
            aria-label="Delete font-size-lg"
            class="icon-button visibility-hidden"
            role="button"
            tabindex="0"
          >
            <div
              class="icon icon--trash"
            />
          </div>
        </div>
      </div>
    </div>
    <div
      class=""
      id="modal"
    >
      <div
        class="modal-dialog"
      >
        <div
          class="modal-dialog-header"
        >
          <h2
            class="section-title"
          >
            Font Size
          </h2>
          <div
            aria-label="Close"
            class="icon-button"
            role="button"
            tabindex="0"
          >
            <div
              class="icon icon--close"
            />
          </div>
        </div>
        <div
          class="modal-dialog-body"
        >
          <form>
            <div
              class="input mb-sm form-group"
            >
              <input
                class="input__field"
                id="name"
                placeholder="font-size-base"
                type="text"
                value=""
              />
              <p
                class="visibility-hidden"
              />
            </div>
            <div
              class="input mb-sm form-group"
            >
              <input
                class="input__field"
                id="value"
                maxlength="6"
                placeholder="16"
                type="text"
                value=""
              />
              <p
                class="visibility-hidden"
              />
            </div>
            <div
              class="d-flex justify-content-end"
            >
              <button
                class="button button--primary"
                type="submit"
              >
                Save
              </button>
            </div>
          </form>
        </div>
      </div>
    </div>
  </section>
</DocumentFragment>
`;
